File: data/src/main/java/com.data/server/UserManagerImpl.java
package com.data.server;

import org.hisp.dhis.android.core.D2;
import org.hisp.dhis.android.core.user.User;

import io.reactivex.Observable;
import io.reactivex.annotations.NonNull;
import retrofit2.Response;

class UserManagerImpl implements UserManager {
    private final D2 d2;

    UserManagerImpl(@NonNull D2 d2) {
        this.d2 = d2;
    }

    @NonNull
    @Override
    public Observable<Response<User>> logIn(@NonNull String username, @NonNull String password) {
        return Observable.defer(() -> Observable.fromCallable(d2.logIn(username, password)));
    }

    @NonNull
    @Override
    public Observable<Boolean> isUserLoggedIn() {
        return Observable.defer(() -> Observable.fromCallable(d2.isUserLoggedIn()));
    }
}

File: stock-managment/src/main/java/org/dhis2/android/rtsm/services/UserManagerImpl.kt
package org.dhis2.android.rtsm.services

import android.content.Intent
import io.reactivex.Completable
import io.reactivex.Observable
import io.reactivex.Single
import org.hisp.dhis.android.core.D2
import org.hisp.dhis.android.core.user.User
import org.hisp.dhis.android.core.user.openid.IntentWithRequestCode
import org.hisp.dhis.android.core.user.openid.OpenIDConnectConfig
import javax.inject.Inject

class UserManagerImpl @Inject constructor(val d2: D2) : UserManager {

    override fun login(username: String, password: String, serverUrl: String): Observable<User?> {
        return Observable.defer {
            d2.userModule().logIn(username, password, serverUrl).toObservable()
        }
    }

    override fun login(config: OpenIDConnectConfig): Observable<IntentWithRequestCode?> {
        return Observable.defer {
            d2.userModule().openIdHandler().logIn(config).toObservable()
        }
    }

    override fun handleAuthData(
        serverUrl: String,
        data: Intent?,
        requestCode: Int
    ): Observable<User?> {
        return Observable.defer {
            d2.userModule().openIdHandler()
                .handleLogInResponse(serverUrl, data, requestCode).toObservable()
        }
    }

    override fun isUserLoggedIn(): Observable<Boolean?> {
        return Observable.defer {
            d2.userModule().isLogged.toObservable()
        }
    }

    override fun userName(): Single<String?> {
        return Single.defer{
            d2.userModule().userCredentials().get().map { it.username() }
        }
    }

    override fun logout(): Completable? {
        return d2.userModule().logOut()
    }
}